src/segment-loader.js

```diff
@@ -154,6 +154,159 @@ export const timestampOffsetForSegment = ({
+ * segment until that value is provided.
+ *
+ * CASE 2: Video prepares to load a new timeline before audio:
+ *
+ * Timeline:       0                 1
+ * Audio Segments: 0 1 2 3 4 5 DISCO 6 7 8 9
+ * Audio Loader:             ^
+ * Video Segments: 0 1 2 3 4 5 DISCO 6 7 8 9
+ * Video Loader                      ^
+ *
+ * In the above example, the video loader is preparing to load the 6th segment, the first
+ * after a discontinuity, while the audio loader is still loading the 5th segment, before
+ * the discontinuity.
+ *
+ * If the video loader goes ahead and loads and appends the 6th segment, then once the
+ * segment is loaded and processed, both the video and audio timestamp offsets will be
+ * set, since video is used as the primary timing source. This is to ensure content lines
+ * up appropriately, as any modifications to the video timing are reflected by audio when
+ * the video loader sets the audio and video timestamp offsets to the same value. However,
+ * setting the timestamp offset for audio before audio has had a chance to change
+ * timelines will likely lead to desyncing, as the audio loader will append segment 5 with
+ * a timestamp intended to apply to segments from timeline 1 rather than timeline 0.
+ *
+ * CASE 3: When seeking, audio prepares to load a new timeline before video
+ *
+ * Timeline:       0                 1
+ * Audio Segments: 0 1 2 3 4 5 DISCO 6 7 8 9
+ * Audio Loader:           ^
+ * Video Segments: 0 1 2 3 4 5 DISCO 6 7 8 9
+ * Video Loader            ^
+ *
+ * In the above example, both audio and video loaders are loading segments from timeline
+ * 0, but imagine that the seek originated from timeline 1.
+ *
+ * When seeking to a new timeline, the timestamp offset will be set based on the expected
+ * segment start of the loaded video segment. In order to maintain sync, the audio loader
+ * must wait for the video loader to load its segment and update both the audio and video
+ * timestamp offsets before it may load and append its own segment. This is the case
+ * whether the seek results in a mismatched segment request (e.g., the audio loader
+ * chooses to load segment 3 and the video loader chooses to load segment 4) or the
+ * loaders choose to load the same segment index from each playlist, as the segments may
+ * not be aligned perfectly, even for matching segment indexes.
+ *
+ * @param {Object} timelinechangeController
+ * @param {number} currentTimeline
+ *        The timeline currently being followed by the loader
+ * @param {number} segmentTimeline
+ *        The timeline of the segment being loaded
+ * @param {('main'|'audio')} loaderType
+ *        The loader type
+ * @param {boolean} audioDisabled
+ *        Whether the audio is disabled for the loader
+ *
+ * @return {boolean}
+ *         Whether the loader should wait for a timeline change from the timeline change
+ *         controller before processing the segment
+ */
+export const shouldWaitForTimelineChange = ({
```

Comment:
should this method just be on the `timelineChangeController`?
so you could do something like
```
const shouldWait = timelineChangeController.shouldWaitForTimelineChange({
  currentTimeline,
  segmentTimeline,
  loaderType: 'main'
});
```

Reply:
I thought about it a bit, and I think that it would be best to keep it out of the TimelineChangeController. Right now, TimelineChangeController is a pretty simple class without any logic in it around loaders or streaming decisions. That logic is primarily contained within the segment loaders, so I figure it is best to keep it contained there. Eventually it might be best to even move this function out of the segment loader module, along with some other logic that ties loaders together, into a separate module, to start shrinking the module itself, but for now I figure it's best to prevent any of the logic leaking.

It also makes testing easier to have it as function not attached to a class.

But I'm also open to hearing arguments for moving it.

Reply:
Main argument for moving it is that this method directly deals with the timeline change and is given the controller. We already have the controller in the call site and the controller already has all the other info, so, it seems reasonable to ask it for whether it should change.
However, it is true, without that method there, the controller does stay simple. Though, the complexity doesn't go away it's just elsewhere.
I'm OK with just leaving as is.